plugin/hosts/hostsfile.go

```diff
@@ -52,71 +44,79 @@ type Hostsfile struct {
 	// including IPv6 address with zone identifier.
 	// We don't support old-classful IP address notation.
 	byAddr map[string][]string
+}
 
-	// inline saves the hosts file is inlined in Corefile
-	// We need a copy here as we want to use inline to override
-	// the default /etc/hosts
-	inline []string
+func newHostsMap() *hostsMap {
+	return &hostsMap{
+		byNameV4: make(map[string][]net.IP),
+		byNameV6: make(map[string][]net.IP),
+		byAddr:   make(map[string][]string),
+	}
+}
 
-	expire time.Time
-	path   string
-	mtime  time.Time
-	size   int64
+// Hostsfile contains known host entries.
+type Hostsfile struct {
+	sync.RWMutex
+
+	// list of zones we are authoritive for
+	Origins []string
+
+	// hosts maps for lookups
+	hmap *hostsMap
+
+	// inline saves the hosts file that is inlined in a Corefile.
+	// We need a copy here as we want to use it to initialize the maps for parse.
+	inline *hostsMap
+
+	// path to the hosts file
+	path string
+
+	// mtime and size are only read and modified by a single goroutine
+	mtime time.Time
+	size  int64
 }
 
 // ReadHosts determines if the cached data needs to be updated based on the size and modification time of the hostsfile.
 func (h *Hostsfile) ReadHosts() {
-	now := time.Now()
-
-	// Not deferring h.RUnlock() because we may need to remove the read lock and aquire a write lock
-	h.RLock()
-	if now.Before(h.expire) && len(h.byAddr) > 0 {
-		h.RUnlock()
+	file, err := os.Open(h.path)
+	if err != nil {
 		return
 	}
-	stat, err := os.Stat(h.path)
+	defer file.Close()
+
+	stat, err := file.Stat()
 	if err == nil && h.mtime.Equal(stat.ModTime()) && h.size == stat.Size() {
-		h.RUnlock()
-		h.Lock()
-		h.expire = now.Add(cacheMaxAge)
-		h.Unlock()
 		return
 	}
 
-	var file *os.File
-	if file, _ = os.Open(h.path); file == nil {
-		// If this is the first time then we will try to parse inline
-		if len(h.byAddr) == 0 && len(h.inline) > 0 {
-			h.Parse(nil)
-		}
-		h.RUnlock()
-		return
-	}
-	defer file.Close()
-
-	h.RUnlock()
 	h.Lock()
 	defer h.Unlock()
-	h.Parse(file)
+	h.parseFile(file)
 
 	// Update the data cache.
-	h.expire = now.Add(cacheMaxAge)
 	h.mtime = stat.ModTime()
 	h.size = stat.Size()
 }
 
-// Parse reads the hostsfile and populates the byName and byAddr maps.
-func (h *Hostsfile) Parse(file io.Reader) {
-	hsv4 := make(map[string][]net.IP)
-	hsv6 := make(map[string][]net.IP)
-	is := make(map[string][]string)
-
-	var readers []io.Reader
-	if file != nil {
-		readers = append(readers, file)
+func (h *Hostsfile) initInline(inline []string) {
+	if len(inline) == 0 {
+		return
 	}
-	readers = append(readers, strings.NewReader(strings.Join(h.inline, "\n")))
-	scanner := bufio.NewScanner(io.MultiReader(readers...))
+
+	hmap := newHostsMap()
+	h.inline = h.parse(strings.NewReader(strings.Join(inline, "\n")), hmap)
+	*h.hmap = *h.inline
+}
+
+func (h *Hostsfile) parseFile(file io.Reader) {
```

Comment:
`parseFile` -> `parseReader`? And just use `r` for the io.Reader instead of `file`?